packages/evm/src/stack.ts

```diff
@@ -49,15 +41,22 @@ export class Stack {
    * @param num - Number of items to pop
    */
   popN(num: number = 1): bigint[] {
-    if (this._store.length < num) {
+    if (this._len < num) {
       throw new EvmError(ERROR.STACK_UNDERFLOW)
     }
 
     if (num === 0) {
       return []
     }
 
-    return this._store.splice(-1 * num).reverse()
+    const arr = Array(num)
+    const cache = this._store
+
+    for (let pop = 0; pop < num; pop++) {
+      arr[pop] = cache[--this._len]
```

Comment:
Interesting, I have not checked, I honestly assumed that this would be faster instead of using `slice`. This should definitely be checked :)

Reply:
I tested: 

```typescript
  popN(num: number = 1): bigint[] {
    if (this._len < num) {
      throw new EvmError(ERROR.STACK_UNDERFLOW)
    }

    if (num === 0) {
      return []
    }

    const arr = this._store.slice(this._len - num, this._len).reverse()

    this._len -= num

    return arr
  }
  ```
  Results
  
  ```
  popN 1 on stack size 1: Diff (neg is better): 0.249s (38.97%)
popN 1 on full-stack: Diff (neg is better): 0.279s (41.94%)
popN 2 on stack size 2: Diff (neg is better): 0.346s (50.15%)
popN 2 on full-stack: Diff (neg is better): 0.337s (49.35%)
popN 3 on stack size 3: Diff (neg is better): 0.313s (42.46%)
popN 3 on full-stack: Diff (neg is better): 0.27s (34.98%)
popN 4 on stack size 4: Diff (neg is better): 0.299s (37.97%)
popN 4 on full-stack: Diff (neg is better): 0.282s (36.32%)
popN 5 on stack size 5: Diff (neg is better): 0.256s (31.47%)
popN 5 on full-stack: Diff (neg is better): 0.25s (30.86%)
popN 6 on stack size 6: Diff (neg is better): 0.252s (29.89%)
popN 6 on full-stack: Diff (neg is better): 0.238s (27.67%)
popN 7 on stack size 7: Diff (neg is better): 0.257s (29.06%)
popN 7 on full-stack: Diff (neg is better): 0.227s (26%)
popN 8 on stack size 8: Diff (neg is better): 0.214s (23.62%)
popN 8 on full-stack: Diff (neg is better): 0.208s (23%)
```

It is thus slower, but the "slowness" seems to decrease, so for larger arrays it might indeed be faster. But it seems not for small arrays?

Reply:
Just to be sure I'm tracking, the `diff` is the time difference between running it with your original optimized loop versus `.slice.reverse()`  and the slice.reverse version is .25s slower?

Reply:
Yes

Reply:
Cool, thanks for exploring this.